### ISSUE TITLE
meshio for gmsh i/o

### ISSUE THREAD

**Issue by nschloe** [usnistgov/fipy#571]
The author of [meshio](https://github.com/nschloe/meshio) here.

I noticed that in https://github.com/usnistgov/fipy/blob/develop/fipy/meshes/gmshMesh.py there's quite a bit of code dealing with gmsh i/o. Much of it can probably be replaced by a meshio one-liner.

**Comment by wd15** [comment]
@nschloe, thanks for pointing this out. What a fantastic project! It's exactly what the community needs. It's great that it is both command line and has a Python API.

Another need is an interactive python-based open-source mesher. Maybe a wrapper for Gmsh. Any thoughts about that?

Unfortunately, we don't have much man power to do things to FiPy these days, but I'll definitely keep this in mind and make it a priority when I revisit FiPy.

**Comment by nschloe** [comment]
>  Maybe a wrapper for Gmsh. 

A wrapper for gmsh is [pygmsh](https://github.com/nschloe/pygmsh), but it's not "interactive". (I'm assuming by that you mean a GUI where you can add geometric entities and such.) If you come across something new, I could add it to https://github.com/nschloe/awesome-numerical-software for anyone else to find.

**Comment by wd15** [comment]
On Fri, May 18, 2018 at 1:17 PM, Nico Schlömer <notifications@github.com> wrote:
>
> Maybe a wrapper for Gmsh.
>
> A wrapper for gmsh is pygmsh, but it's not "interactive". (I'm assuming by that you mean a GUI where you can add geometric entities and such.)

Actually, no. Interactive at the command prompt is what I meant, but
didn't say. It would be nice to be in the Jupyter Notebook and be able
to do

    >>> mesh = makeMyMesh(domainDescription, meshingControlParameters)
    >>> mesh.vertices

Basically, build up parts of a mesh at the Python prompt, add them
together, delete. View it in Jupyter. Interactively change the mesh.


-- 
Daniel Wheeler

**Comment by nschloe** [comment]
Yup, then [pygmsh](https://github.com/nschloe/pygmsh) is what you're looking for. Check out the example in the readme.

**Comment by wd15** [comment]
On Fri, May 18, 2018 at 2:51 PM, Nico Schlömer <notifications@github.com> wrote:
>
> Yup, then pygmsh is what you're looking for. Check out the example in the readme.

Wow! Really appreciate that you made us aware of these great
libraries. FiPy should definitely integrate with both of these.

-- 
Daniel Wheeler

**Comment by guyer** [comment]
I second @wd15: These packages are fantastic. If I can ever get our CI running again, I'd like to dive in and integrate these. 

How do your packages behave when running in parallel?

**Comment by nschloe** [comment]
What exactly do you mean by "parallel"? [multiprocessing](https://docs.python.org/3/library/multiprocessing.html)?

**Comment by guyer** [comment]
When run under MPI, FiPy directs gmsh to partition the mesh and then loads each of those mesh partitions onto each of the separate processes. To write a gmsh, Vtk, XDMF or whatever, we'd need to consolidate that information from the processes into a single file (or set of them, where feasible). Do pygmsh and meshio help with those tasks?

**Comment by nschloe** [comment]
Nothing like that is implemented in pygmsh.
